doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,209 @@
+# ADR 5: Plutus Tx's Evaluation Strategy, Compiler and Standard Library
+
+Date: 2023-04
+
+## Authors
+
+Ziyang Liu <ziyang.liu@iohk.io>
+
+## Status
+
+Proposed
+
+## Background
+
+Plutus Tx is a language for writing Plutus scripts.
+It is a subset of Haskell and can be compiled to Untyped Plutus Core (hereafter referred to as PLC) by the Plutus Tx compiler, a GHC plugin.
+
+Haskell uses lazy evaluation (call by need) by default, while PLC is strict (call by value).
+This begs the question: what is the operational semantics of Plutus Tx?
+This is an important question, the answer of which not only dictates how users of Plutus Tx should reason about and optimize their programs, but also has major implications on the design and implementation of the Plutus Tx compiler and libraries.
+
+Unfortunately, this question is currently difficult to answer.
+One thing is clear: Plutus Tx certainly doesn't have lazy semantics: PLC's evaluator (the CEK machine) has no support for thunks and memoization.
+What is unclear is whether Plutus Tx is strict or non-strict.
+Indeed, it is sometimes strict and sometimes non-strict, and it is fairly random and difficult to predict:
+
+- Let-bindings in Plutus Tx are by default non-strict, but function applications are strict.
+  That is, `let x = rhs in body` isn't equivalent to `(\x -> body) rhs`.
+  This is distinct from most other languages, regardless of their evaluation strategies, in which these two expressions are equivalent.
+
+- The conditional expression `if b then t else f` is always non-strict in `t` and `f`, like in most strict languages.
+  But what about `foo b t f = if b then t else f`? In a strict language, `foo` would be strict, and in a non-strict language, it would be non-strict.
+  But in Plutus Tx, due to how its compiler is currently implemented, whether `t` and `f` are evaluated strictly or not depends on whether GHC inlines `foo`.
+  Although we turn off GHC's inliner in the Plutus Tx compiler, GHC may still decide to unconditionally inline `foo` if it is used only once.
+  In general, it is not easy to predict whether GHC will inline something or not.
+
+The lack of consistency makes it difficult to reason about the efficiency of Plutus Tx programs, and difficult to write useful libraries for Plutus Tx that guarantee to have the desirable behaviors (e.g., short-circuiting).
+
+This ADR proposes a few changes to the Plutus Tx language and the compiler to make the semantics more predictable.
+It also proposes changes to the Plutus Tx standard library to make it more performant.
+
+## Proposed Changes
+
+Plutus Tx should be a strict-by-default language.
+Non-strict-by-default languages are rarely used in practice since non-strict evaluation leads to repeated evaluations.
```

Comment:
I think there's a lot of terminology confusion here.

1. strict vs non-strict isn't about evaluation per se, that distinction lies within the realms of _denotational_ semantics. E.g. `seq` is strict in both of its arguments, which doesn't tell us much about how things _evaluate_
2. call-by-value is an evaluation strategy than a strict language can adopt. Similarly, call-by-name or call-by-need are evaluation strategies that a non-strict language can adopt. At this level we're talking about the _operational_ semantics of a language. For `seq` we'd say that it's supposed to evaluate its second argument after evaluating the first one. However if GHC devs were to make `seq` evaluate both of its arguments in parallel, that would still comply with the denotational specification of the function
3. eager vs lazy is fairly vague terms for describing the specific _implementation_ of the evaluation strategy of a language

As such, non-strictness doesn't lead to any repeated evaluation as it isn't about evaluation at all. Laziness can lead to repeated evaluation if the call-by-name evaluation strategy is chosen, but that's horrible enough for it to be very uncommon (Agda used to be call-by-name for an example, not sure how want they've gone in making it call-by-need instead). In practice lazy languages implement call-by-need (Haskell, R) and that doesn't lead to repeated evaluation.